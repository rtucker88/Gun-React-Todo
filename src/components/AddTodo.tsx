import * as React from "react";
import { InputGroup, Button } from "@blueprintjs/core";
import { compose, withHandlers, withStateHandlers, setDisplayName, StateHandlerMap, StateHandler } from "recompose";

import "../assets/scss/AddTodo.scss";
import { translate } from "react-i18next";
import { TransProps } from "react-i18next/src/trans";

export interface AddTodoProps {
    onAddTodo: (text: string) => void;
}

export interface AddTodoState {
    text: string;
}

export interface AddTodoStateHandlers extends StateHandlerMap<AddTodoState> {
    onInputChange: StateHandler<AddTodoState>;
    onAddNewTodo: StateHandler<AddTodoState>;
}

interface AddTodoHandlers {
    onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
}

export const stateHandlers = withStateHandlers<AddTodoState, AddTodoStateHandlers, AddTodoProps>({ text: "" }, {
    onInputChange: () => (event: React.ChangeEvent<HTMLInputElement>) => {
        return {
            text: event.target.value,
        };
    },
    onAddNewTodo: (state, props) => () => {
        props.onAddTodo(state.text);

        return {
            text: "",
        };
    },
});

export const handlers = withHandlers<AddTodoProps & AddTodoStateHandlers, AddTodoHandlers>({
    onKeyDown: (props: AddTodoCombinedProps) => (event: React.KeyboardEvent<HTMLInputElement>) => {
        if (event.keyCode === 13) {
            props.onAddNewTodo();
        }
    }
});

const enhance = compose<AddTodoCombinedProps, AddTodoProps>(setDisplayName("AddTodo"), translate("addTodos"), stateHandlers, handlers);

export type AddTodoCombinedProps = AddTodoState & AddTodoStateHandlers & AddTodoHandlers & TransProps & AddTodoProps;

export const AddTodo: React.StatelessComponent<AddTodoCombinedProps> = ({ onAddNewTodo, onKeyDown, onInputChange, t, text }) => {
    return (
        <div className="add-todo-container">
            <div className="add-todo">
                <InputGroup onKeyDown={onKeyDown} onChange={onInputChange} value={text} className="add-todo-input" placeholder={t("enterANewTodo")}/>
                <Button onClick={onAddNewTodo} icon="add" minimal={true} />
            </div>
        </div>);
};

export default enhance(AddTodo);
