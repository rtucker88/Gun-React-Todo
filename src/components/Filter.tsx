import * as React from "react";
import { ButtonGroup, Button } from "@blueprintjs/core";

import "../assets/scss/Filter.scss";
import { TransProps } from "react-i18next/src/trans";
import { translate } from "react-i18next";
import { compose, setDisplayName, withHandlers } from "recompose";

interface FilterProps {
    filterToggle: FilterToggle;
    onUpdateFilter: (toggle: FilterToggle) => void;
}

interface FilterHandlerProps {
    updateAll: () => void;
    updateIncomplete: () => void;
    updateCompleted: () => void;
}

export type FilterCombinedProps = FilterProps & TransProps & FilterHandlerProps;

export const handlers = withHandlers<FilterProps, FilterHandlerProps>({
    updateAll: props => () => props.onUpdateFilter("ALL"),
    updateIncomplete: props => () => props.onUpdateFilter("INCOMPLETE"),
    updateCompleted: props => () => props.onUpdateFilter("COMPLETED"),
});

const enhance = compose<{}, FilterProps>(setDisplayName("Filter"), translate("filterToggle"), handlers);

export type FilterToggle = "ALL" | "COMPLETED" | "INCOMPLETE";

const Filter: React.SFC<FilterCombinedProps> = ({ filterToggle, updateAll, updateCompleted, updateIncomplete, t }) => {
    return (
    <div className="filter">
        <ButtonGroup>
            <Button active={filterToggle === "ALL"} onClick={updateAll}>{t("all")}</Button>
            <Button active={filterToggle === "INCOMPLETE"} onClick={updateIncomplete} icon="square">{t("incomplete")}</Button>
            <Button active={filterToggle === "COMPLETED"} onClick={updateCompleted} icon="tick">{t("completed")}</Button>
        </ButtonGroup>
    </div>);
};

export default enhance(Filter);