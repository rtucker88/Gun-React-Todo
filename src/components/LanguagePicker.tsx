import * as React from "react";
import { Button, Popover, MenuItem, Menu, Position } from "@blueprintjs/core";
import UnreachableCaseError from "../utils/UnreachableErrorCase";
import { setDisplayName, pure, compose } from "recompose";

interface LanguagePickerProps {
    onChangeLanguage: (language: I18nLanguage) => void;
    currentLanguage: I18nLanguage;
}

export type I18nLanguage = "en" | "de" | "el" | "uk";

interface MenuItemConfig {
    country: string;
    languageName: string;
    onClickParameter: I18nLanguage;
}

const menuItems: MenuItemConfig[] = [{
    country: "US",
    languageName: "English",
    onClickParameter: "en"
},
{
    country: "DE",
    languageName: "Deutsch",
    onClickParameter: "de"
},
{
    country: "GR",
    languageName: "ελληνικά",
    onClickParameter: "el"
},
{
    country: "UA",
    languageName: "Українська",
    onClickParameter: "uk"
}];

const LanguagePicker: React.StatelessComponent<LanguagePickerProps> = ({ currentLanguage, onChangeLanguage }) => {
    return (
        <Popover content={getMenuItems(onChangeLanguage)} position={Position.BOTTOM_LEFT}>
            <Button icon="caret-down" text={getCurrentLocaleFlag(currentLanguage)} minimal={true}></Button>
        </Popover>
    );
};

const enhance = compose<{}, LanguagePickerProps>(setDisplayName("LanguagePicker"), pure);
export default enhance(LanguagePicker);

export function getMenuItems(onClick: (language: I18nLanguage) => void) {
    return (<Menu>
        {menuItems.map(({ country, languageName, onClickParameter }) => {
            return (<MenuItem key={country} text={`${getFlagFromCountryCode(country)} ${languageName}`} onClick={() => onClick(onClickParameter)} />);
        })}
    </Menu>);
}


export function getCurrentLocaleFlag(locale: I18nLanguage): string {
    switch (locale) {
        case "en":
            return getFlagFromCountryCode("US");
        case "de":
            return getFlagFromCountryCode("DE");
        case "el":
            return getFlagFromCountryCode("GR");
        case "uk":
            return getFlagFromCountryCode("UA");
        default:
            throw new UnreachableCaseError(locale);
    }
}

export function getFlagFromCountryCode(countryCode: string): string {
    // Taken from https://binarypassion.net/lets-turn-an-iso-country-code-into-a-unicode-emoji-shall-we-870c16e05aad
    return countryCode.toUpperCase().replace(/./g, char => String.fromCodePoint(char.charCodeAt(0) + 127397));
}