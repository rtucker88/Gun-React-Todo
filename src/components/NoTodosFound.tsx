import * as React from "react";
import { NonIdealState } from "@blueprintjs/core";

import "../assets/scss/NoTodosFound.scss";
import { translate } from "react-i18next";
import { TransProps } from "react-i18next/src/trans";
import { setDisplayName, compose, pure } from "recompose";

interface NoTodosFoundProps {}

type NoTodosFoundCombinedProps = NoTodosFoundProps & TransProps;

const NoTodosFound: React.SFC<NoTodosFoundCombinedProps> = ({ t }) => {
    return (<div className="no-todos-found">
        <NonIdealState title={t("noneFoundTitle")} description={t("noneFoundDescription")} icon="zoom-out" />
    </div>);
};

const enhance = compose(setDisplayName("NoTodosFound"), pure, translate("noTodosFound"));

export default enhance(NoTodosFound);
