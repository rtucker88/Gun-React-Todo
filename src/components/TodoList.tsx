import * as React from "react";
import TodoItem from "./Todo";

import "../assets/scss/TodoList.scss";
import { compose, setDisplayName, pure } from "recompose";

export interface Todo {
    contents: string;
    completed: boolean;
    timeCreated: number;
    id: string;
}

interface TodoListProps {
    todos: Todo[];
    onDelete: (id: string) => void;
    onComplete: (id: string) => void;
}

const TodoList: React.SFC<TodoListProps> = ({ onComplete, onDelete, todos }) => {
    return (<div className="todo-list">
        {todos.map(({ completed, contents, id, timeCreated }) => <TodoItem completed={completed} contents={contents} key={`${timeCreated}-${contents}`} onDelete={onDelete} onComplete={onComplete} id={id} timeCreated={timeCreated} />)}
    </div>);
};

const enhance = compose<{}, TodoListProps>(setDisplayName("TodoList"), pure);

export default enhance(TodoList);
