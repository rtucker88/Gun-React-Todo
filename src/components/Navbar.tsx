import * as React from "react";
import { Navbar as BlueprintNavbar, NavbarGroup, NavbarHeading, NavbarDivider, Button, Classes, Alignment } from "@blueprintjs/core";
import { default as LanguagePicker, I18nLanguage } from "./LanguagePicker";
import { translate } from "react-i18next";
import { TransProps } from "react-i18next/src/trans";
import { compose, setDisplayName, pure } from "recompose";

interface NavbarProps {
    onChangeLanguage: (language: I18nLanguage) => void;
    currentLanguage: I18nLanguage;
}

type CombinedNavbarProps = NavbarProps & TransProps;

const Navbar: React.SFC<CombinedNavbarProps> = ({ onChangeLanguage, currentLanguage, t }) => {
    return (<BlueprintNavbar>
            <NavbarGroup>
                <NavbarHeading>Gun-React-Todo</NavbarHeading>
                <NavbarDivider />
                <Button className={Classes.MINIMAL} icon="home" text={t("home")} />
                <Button className={Classes.MINIMAL} icon="help" text={t("about")} />
            </NavbarGroup>
            <NavbarGroup align={Alignment.RIGHT}>
                <LanguagePicker onChangeLanguage={onChangeLanguage} currentLanguage={currentLanguage} />
            </NavbarGroup>
        </BlueprintNavbar>);
};

const enhance = compose<{}, NavbarProps>(setDisplayName("Navbar"), pure, translate("navbar"));

export default enhance(Navbar);