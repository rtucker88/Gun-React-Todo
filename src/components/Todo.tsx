import * as React from "react";
import classNames from "classnames";
import { withHandlers, compose, setDisplayName } from "recompose";
import { Card, Button, Checkbox } from "@blueprintjs/core";
import { formatRelative } from "date-fns";
import { enUS, de } from "date-fns/esm/locale";

import "../assets/scss/Todo.scss";
import { translate } from "react-i18next";
import { TransProps } from "react-i18next/src/trans";
import { I18nLanguage } from "./LanguagePicker";

interface TodoProps {
    contents: string;
    completed: boolean;
    id: string;
    timeCreated: number;
    onDelete: (id: string) => void;
    onComplete: (id: string) => void;
}

interface TodoHandlerProps {
    handleComplete: () => void;
    handleDelete: () => void;
}

export const handlers = withHandlers<TodoProps, TodoHandlerProps>({
    handleComplete: props => () => props.onComplete(props.id),
    handleDelete: props => () => props.onDelete(props.id),
});

const enhance = compose<TodoHandlerProps, TodoProps>(setDisplayName("Todo"), translate("todo"), handlers);

export type CombinedTodoProps = TodoProps & TodoHandlerProps & TransProps;

const Todo: React.SFC<CombinedTodoProps> = ({ completed, contents, i18n, handleComplete, handleDelete, t, timeCreated }) => {
    return (<Card className={classNames("todo", { completed })}>
            <Checkbox checked={completed} onChange={handleComplete} className="aligner"/>
            <p className="aligner">{contents}</p>
            <Button icon="trash" onClick={handleDelete} minimal={true} intent="danger" className="aligner" />
            <span className="created-date">{`${t("created")} ${formatRelative(timeCreated, Date.now(), { locale: getDateFnLocale((i18n as any).default.language as I18nLanguage) })}`}</span>
    </Card>);
};


export default enhance(Todo);

export function getDateFnLocale(language: I18nLanguage) {
    switch (language) {
        case "en":
            return enUS;
        case "de":
            return de;
        case "el":
        case "uk":
            // Greek and Ukranian aren't supported in the newest version, update this later
            return enUS;
        default:
            return enUS;
    }
}