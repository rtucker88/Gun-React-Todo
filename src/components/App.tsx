import * as React from "react";
import * as uuidv4 from "uuid/v4";
import * as i18nnext from "i18next";
import Navbar from "./Navbar";
import { TodoList, Todo } from "./TodoList";
import AddTodo from "./AddTodo";
import { default as Filter, FilterToggle } from "./Filter";
import UnreachableCaseError from "../utils/UnreachableErrorCase";
import NoTodosFound from "./NoTodosFound";

import "../assets/scss/App.scss";
import { I18nLanguage } from "./LanguagePicker";
import i18n from "../i18n";

export interface AppProps {
}

interface AppState {
    todos: Todo[];
    filterToggle: FilterToggle;
    currentLanguage: I18nLanguage;
}

export default class App extends React.Component<AppProps, AppState> {
    state = {
        todos: [],
        filterToggle: "ALL" as FilterToggle,
        currentLanguage: "en" as I18nLanguage,
    };

    componentWillMount() {
        // Trim this here because we'll get things like en-US and en-UK
        this.setState({
            currentLanguage: (i18nnext as any).default.language.substr(0, 2) as I18nLanguage,
        });
    }

    private static getFilteredTodos(todos: Todo[], toggle: FilterToggle): Todo[] {
        return todos.filter(todo => {
            switch (toggle) {
                case "ALL":
                    return true;
                case "COMPLETED":
                    return todo.completed;
                case "INCOMPLETE":
                    return !todo.completed;
                default:
                    throw new UnreachableCaseError(toggle);
            }
        });
    }

    private handleAddTodo = (text: string) => {
        this.setState((state: Readonly<AppState>) => {
            return {
                todos: [...state.todos, { contents: text, completed: false, timeCreated: Date.now(), id: uuidv4() }],
            };
        });
    }

    private handleTodoCompletion = (id: string) => {
        this.setState((state: Readonly<AppState>) => {
            const todo = state.todos.find(todo => todo.id === id);
            todo.completed = !todo.completed;

            return {
                todos: state.todos,
            };
        });
    }

    private handleTodoDelete = (id: string) => {
        this.setState((state: Readonly<AppState>) => {
            const todoIndex = state.todos.findIndex(todo => todo.id === id);
            state.todos.splice(todoIndex, 1);

            return {
                todos: state.todos,
            };
        });
    }

    private handleUpdateFilter = (filterToggle: FilterToggle) => {
        this.setState({
            filterToggle,
        });
    }

    private handleLanguageChange = (currentLanguage: I18nLanguage) => {
        i18n.changeLanguage(currentLanguage);

        this.setState({
            currentLanguage,
        });
    }

    render() {
        const { currentLanguage, filterToggle, todos } = this.state;

        const filteredTodos = App.getFilteredTodos(todos, filterToggle);

        return (
            <div className="app">
                <Navbar currentLanguage={currentLanguage} onChangeLanguage={this.handleLanguageChange}></Navbar>
                <h1 className="todo-header">Todos</h1>
                <AddTodo onAddTodo={this.handleAddTodo} />
                <Filter filterToggle={filterToggle} onUpdateFilter={this.handleUpdateFilter} />
                {filteredTodos.length ? <TodoList todos={filteredTodos} onComplete={this.handleTodoCompletion} onDelete={this.handleTodoDelete} /> : <NoTodosFound />}
            </div>
        );
    }
}
