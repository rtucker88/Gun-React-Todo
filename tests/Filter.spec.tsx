import * as React from "react";
import { mount } from "enzyme";
import { default as Filter, FilterCombinedProps, handlers } from "../src/components/Filter";

describe("Filter", () => {
   class MockComponent extends React.Component<FilterCombinedProps> {
        render() {
            return <div />;
        }
    }

    const onUpdateFilterMock = jest.fn();

    beforeEach(() => {
        onUpdateFilterMock.mockClear();
    });

    it("renders correctly", () => {
        const wrapper = mount((<Filter
            filterToggle="ALL"
            onUpdateFilter={onUpdateFilterMock} />));
        expect(wrapper).toMatchSnapshot();
    });

    describe("handlers", () => {
        it("updates all", () => {
            const Enhanced = handlers(MockComponent);
            const wrapper = mount(<Enhanced onUpdateFilter={onUpdateFilterMock} filterToggle="ALL" />);
            const subject = wrapper.find(MockComponent);
            subject.prop("updateAll")();
            expect(onUpdateFilterMock.mock.calls[0][0]).toEqual("ALL");
        });

        it("updates incomplete", () => {
            const Enhanced = handlers(MockComponent);
            const wrapper = mount(<Enhanced onUpdateFilter={onUpdateFilterMock} filterToggle="ALL" />);
            const subject = wrapper.find(MockComponent);
            subject.prop("updateIncomplete")();
            expect(onUpdateFilterMock.mock.calls[0][0]).toEqual("INCOMPLETE");
        });

        it("updates complete", () => {
            const Enhanced = handlers(MockComponent);
            const wrapper = mount(<Enhanced onUpdateFilter={onUpdateFilterMock} filterToggle="ALL" />);
            const subject = wrapper.find(MockComponent);
            subject.prop("updateCompleted")();
            expect(onUpdateFilterMock.mock.calls[0][0]).toEqual("COMPLETED");
        });
    });
});
