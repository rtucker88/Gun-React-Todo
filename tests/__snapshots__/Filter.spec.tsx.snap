// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter renders correctly 1`] = `
<Filter
  filterToggle="ALL"
  onUpdateFilter={[MockFunction]}
>
  <withHandlers(Filter)
    filterToggle="ALL"
    i18n={
      Object {
        "default": Object {
          "language": "en",
        },
      }
    }
    onUpdateFilter={[MockFunction]}
    t={[Function]}
  >
    <Filter
      filterToggle="ALL"
      i18n={
        Object {
          "default": Object {
            "language": "en",
          },
        }
      }
      onUpdateFilter={[MockFunction]}
      t={[Function]}
      updateAll={[Function]}
      updateCompleted={[Function]}
      updateIncomplete={[Function]}
    >
      <div
        className="filter"
      >
        <Blueprint3.ButtonGroup>
          <div
            className="bp3-button-group"
          >
            <Blueprint3.Button
              active={true}
              onClick={[Function]}
            >
              <button
                className="bp3-button bp3-active"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <Blueprint3.Icon
                  key="leftIcon"
                />
                <span
                  className="bp3-button-text"
                  key="text"
                >
                  all
                </span>
                <Blueprint3.Icon
                  key="rightIcon"
                />
              </button>
            </Blueprint3.Button>
            <Blueprint3.Button
              active={false}
              icon="square"
              onClick={[Function]}
            >
              <button
                className="bp3-button"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <Blueprint3.Icon
                  icon="square"
                  key="leftIcon"
                >
                  <span
                    className="bp3-icon bp3-icon-square"
                    icon="square"
                  >
                    <svg
                      data-icon="square"
                      height={16}
                      viewBox="0 0 16 16"
                      width={16}
                    >
                      <desc>
                        square
                      </desc>
                      <path
                        d="M15 0H1C.45 0 0 .45 0 1v14c0 .55.45 1 1 1h14c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1zm-1 14H2V2h12v12z"
                        fillRule="evenodd"
                        key="0"
                      />
                    </svg>
                  </span>
                </Blueprint3.Icon>
                <span
                  className="bp3-button-text"
                  key="text"
                >
                  incomplete
                </span>
                <Blueprint3.Icon
                  key="rightIcon"
                />
              </button>
            </Blueprint3.Button>
            <Blueprint3.Button
              active={false}
              icon="tick"
              onClick={[Function]}
            >
              <button
                className="bp3-button"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <Blueprint3.Icon
                  icon="tick"
                  key="leftIcon"
                >
                  <span
                    className="bp3-icon bp3-icon-tick"
                    icon="tick"
                  >
                    <svg
                      data-icon="tick"
                      height={16}
                      viewBox="0 0 16 16"
                      width={16}
                    >
                      <desc>
                        tick
                      </desc>
                      <path
                        d="M14 3c-.28 0-.53.11-.71.29L6 10.59l-3.29-3.3a1.003 1.003 0 0 0-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l8-8A1.003 1.003 0 0 0 14 3z"
                        fillRule="evenodd"
                        key="0"
                      />
                    </svg>
                  </span>
                </Blueprint3.Icon>
                <span
                  className="bp3-button-text"
                  key="text"
                >
                  completed
                </span>
                <Blueprint3.Icon
                  key="rightIcon"
                />
              </button>
            </Blueprint3.Button>
          </div>
        </Blueprint3.ButtonGroup>
      </div>
    </Filter>
  </withHandlers(Filter)>
</Filter>
`;
